File: server/internal/plansig/plansig_test.go
package plansig

import (
	"bytes"
	"crypto/ed25519"
	"encoding/base64"
	"encoding/hex"
	"errors"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

const plan = "ib-plan\t1\nrecord\ttjfq5rqwnnrxk3m9q2x7v4p8ab\nname\tHello\n\n[target]\nwhen\tlinux\t0\t9999\t*\nlaunch\techo hi\n"

func newSigner(t *testing.T) (*Signer, string) {
	t.Helper()
	dir := t.TempDir()
	s, created, err := LoadOrCreate(dir)
	if err != nil || !created {
		t.Fatalf("create: %v %v", created, err)
	}
	return s, dir
}

func TestKeyFiles(t *testing.T) {
	s, dir := newSigner(t)
	st, err := os.Stat(filepath.Join(dir, KeyFile))
	if err != nil || st.Mode().Perm() != 0o600 {
		t.Fatalf("private key mode: %v %v", st.Mode().Perm(), err)
	}
	pub, _ := os.ReadFile(filepath.Join(dir, PubFile))
	if string(pub) != s.PublicBase64()+"\n" || len(s.PublicBase64()) != 44 {
		t.Fatalf("public key file %q", pub)
	}
	// A second start loads the same key and doesn't make a new one.
	s2, created, err := LoadOrCreate(dir)
	if err != nil || created || !bytes.Equal(s2.Pub, s.Pub) {
		t.Fatalf("reload: %v %v", created, err)
	}
	// The public key file is rewritten from the private key if it drifts.
	os.WriteFile(filepath.Join(dir, PubFile), []byte("junk\n"), 0o644)
	if _, _, err := LoadOrCreate(dir); err != nil {
		t.Fatal(err)
	}
	if pub, _ := os.ReadFile(filepath.Join(dir, PubFile)); string(pub) != s.PublicBase64()+"\n" {
		t.Fatalf("public key not restored: %q", pub)
	}
	if !strings.Contains(s.PublicPEM(), "MCowBQYDK2VwAyEA") {
		t.Fatalf("PEM prefix: %s", s.PublicPEM())
	}
	if len(KeyID(s.Pub)) != 16 {
		t.Fatal("key id")
	}
}

func TestSignVerify(t *testing.T) {
	s, _ := newSigner(t)
	signed, err := s.Sign([]byte(plan))
	if err != nil {
		t.Fatal(err)
	}
	if !bytes.HasPrefix(signed, []byte(plan)) {
		t.Fatal("the plan's bytes changed")
	}
	last := strings.Split(strings.TrimSuffix(string(signed), "\n"), "\n")
	line := last[len(last)-1]
	if !strings.HasPrefix(line, "sig\ted25519\t") || len(line) != len("sig\ted25519\t")+88 || !bytes.HasSuffix(signed, []byte("\n")) {
		t.Fatalf("signature line %q", line)
	}
	// The signature is plain Ed25519 over the plan's bytes: check it
	// without this package, as the engines do.
	sig, _ := base64.StdEncoding.DecodeString(strings.TrimPrefix(line, "sig\ted25519\t"))
	if !ed25519.Verify(s.Pub, []byte(plan), sig) {
		t.Fatal("not a signature over the plan's bytes")
	}
	msg, err := VerifyFor(s.Pub, signed, "tjfq5rqwnnrxk3m9q2x7v4p8ab")
	if err != nil || string(msg) != plan {
		t.Fatalf("verify: %v", err)
	}
	// CRLF on the signature line is tolerated; the signed bytes aren't touched.
	if _, err := Verify(s.Pub, append(bytes.Clone(bytes.TrimSuffix(signed, []byte("\n"))), '\r', '\n')); err != nil {
		t.Fatalf("CRLF: %v", err)
	}
	// So is a missing final newline.
	if _, err := Verify(s.Pub, bytes.TrimSuffix(signed, []byte("\n"))); err != nil {
		t.Fatalf("no final newline: %v", err)
	}
	if _, err := s.Sign(signed); err == nil {
		t.Fatal("signed twice")
	}
	if _, err := s.Sign([]byte("ib-record\t1\n")); err == nil {
		t.Fatal("signed a record")
	}
}

func TestNoFinalNewline(t *testing.T) {
	s, _ := newSigner(t)
	signed, _ := s.Sign([]byte(strings.TrimSuffix(plan, "\n")))
	if !bytes.HasPrefix(signed, []byte(plan)) {
		t.Fatal("the newline before sig must be added (and signed)")
	}
	if _, err := Verify(s.Pub, signed); err != nil {
		t.Fatal(err)
	}
}

func TestRejects(t *testing.T) {
	s, _ := newSigner(t)
	other, _ := newSigner(t)
	signed, _ := s.Sign([]byte(plan))

	cases := map[string][]byte{
		"tampered url":    bytes.Replace(signed, []byte("echo hi"), []byte("echo HI"), 1),
		"dropped line":    bytes.Replace(signed, []byte("name\tHello\n"), nil, 1),
		"added line":      bytes.Replace(signed, []byte("[target]\n"), []byte("[target]\nstep\trun\tevil\n"), 1),
		"text after sig":  append(append([]byte{}, signed...), "launch\tevil\n"...),
		"blank after sig": append(append([]byte{}, signed...), '\n'),
		"flipped sig bit": func() []byte {
			b := append([]byte{}, signed...)
			i := bytes.LastIndex(b, []byte("\t")) + 5
			if b[i] == 'A' {
				b[i] = 'B'
			} else {
				b[i] = 'A'
			}
			return b
		}(),
	}
	for name, doc := range cases {
		if _, err := Verify(s.Pub, doc); err == nil {
			t.Errorf("%s: accepted", name)
		}
	}
	if _, err := Verify(other.Pub, signed); !errors.Is(err, ErrBadSig) {
		t.Errorf("another key: %v", err)
	}
	if _, err := Verify(s.Pub, []byte(plan)); !errors.Is(err, ErrUnsigned) {
		t.Errorf("unsigned: %v", err)
	}
	if _, err := VerifyFor(s.Pub, signed, "aaaaaaaaaaaaaaaaaaaaaaaaaa"); err == nil {
		t.Error("record binding not checked")
	}
	// A validly signed plan for another record is refused: replaying one
	// app's plan for another installer doesn't work.
	p2, _ := s.Sign([]byte(strings.Replace(plan, "tjfq5rqwnnrxk3m9q2x7v4p8ab", "bbbbbbbbbbbbbbbbbbbbbbbbbb", 1)))
	if _, err := VerifyFor(s.Pub, p2, "tjfq5rqwnnrxk3m9q2x7v4p8ab"); err == nil {
		t.Error("replayed plan accepted")
	}
	// Non-canonical S (S + L) is refused.
	sig, _ := base64.StdEncoding.DecodeString(string(signed[bytes.LastIndex(signed, []byte("\t"))+1 : len(signed)-1]))
	addL(sig[32:])
	bad := append(append([]byte{}, signed[:bytes.LastIndex(signed, []byte("\t"))+1]...), base64.StdEncoding.EncodeToString(sig)+"\n"...)
	if _, err := Verify(s.Pub, bad); err == nil {
		t.Error("S >= L accepted")
	}
}

// addL adds the group order L to a little-endian 32-byte scalar.
func addL(s []byte) {
	l, _ := hex.DecodeString("edd3f55c1a631258d69cf7a2def9de1400000000000000000000000000000010")
	c := 0
	for i := 0; i < 32; i++ {
		v := int(s[i]) + int(l[i]) + c
		s[i], c = byte(v), v>>8
	}
}

func TestRecordOf(t *testing.T) {
	if RecordOf([]byte(plan)) != "tjfq5rqwnnrxk3m9q2x7v4p8ab" {
		t.Fatal("record")
	}
	if RecordOf([]byte("ib-plan\t1\n\n[target]\nrecord\tx\n")) != "" {
		t.Fatal("a record line inside a target block is not the header's")
	}
}

File: server/test/plansig.test.js
// Ports the Go server's plansig tests (server/internal/plansig, retired).
import test from 'node:test';
import assert from 'node:assert/strict';
import fs from 'node:fs';
import path from 'node:path';
import crypto from 'node:crypto';
import { loadOrCreate, keyID, split, verify, verifyFor, recordOf, addRequestLine, publicKeyFromRaw, KEY_FILE, PUB_FILE, VerifyError } from '../lib/plansig.js';
import { tmpDir, REPO } from './helpers.js';

const PLAN = 'ti-plan\t1\nrecord\ttjfq5rqwnnrxk3m9q2x7v4p8ab\nname\tHello\n\n[target]\nwhen\tlinux\t0\t9999\t*\nlaunch\techo hi\n';

function newSigner(t) {
  const dir = tmpDir(t);
  const { signer, created } = loadOrCreate(dir, () => {});
  assert.ok(created);
  return { s: signer, dir };
}

test('key files', (t) => {
  const { s, dir } = newSigner(t);
  assert.equal(fs.statSync(path.join(dir, KEY_FILE)).mode & 0o777, 0o600);
  assert.equal(fs.readFileSync(path.join(dir, PUB_FILE), 'utf8'), s.publicBase64() + '\n');
  assert.equal(s.publicBase64().length, 44);
  // A second start loads the same key and doesn't make a new one.
  const again = loadOrCreate(dir, () => {});
  assert.equal(again.created, false);
  assert.ok(again.signer.pub.equals(s.pub));
  // The public key file is rewritten from the private key if it drifts.
  fs.writeFileSync(path.join(dir, PUB_FILE), 'junk\n');
  loadOrCreate(dir, () => {});
  assert.equal(fs.readFileSync(path.join(dir, PUB_FILE), 'utf8'), s.publicBase64() + '\n');
  assert.match(s.publicPEM(), /MCowBQYDK2VwAyEA/);
  assert.equal(keyID(s.pub).length, 16);
});

test('a key readable by others is used, with a warning', (t) => {
  const { dir } = newSigner(t);
  fs.chmodSync(path.join(dir, KEY_FILE), 0o644);
  const logs = [];
  loadOrCreate(dir, (m) => logs.push(m));
  assert.match(logs.join('\n'), /readable by other users/);
});

test('the server\'s key loads and signs deterministically', (t) => {
  const key = path.join(REPO, 'server', 'data', KEY_FILE);
  if (!fs.existsSync(key)) return t.skip('no server/data key');
  const dir = tmpDir(t);
  fs.copyFileSync(key, path.join(dir, KEY_FILE));
  fs.chmodSync(path.join(dir, KEY_FILE), 0o600);
  const { signer, created } = loadOrCreate(dir, () => {});
  assert.equal(created, false);
  assert.equal(fs.readFileSync(path.join(dir, PUB_FILE), 'utf8'), fs.readFileSync(path.join(REPO, 'server', 'data', PUB_FILE), 'utf8'));
  // Ed25519 is deterministic: signing twice gives the same bytes.
  assert.equal(signer.signString(PLAN), signer.signString(PLAN));
});

test('sign and verify', (t) => {
  const { s } = newSigner(t);
  const signed = s.sign(Buffer.from(PLAN));
  assert.ok(signed.subarray(0, PLAN.length).equals(Buffer.from(PLAN)), 'the plan\'s bytes changed');
  const lines = signed.toString().replace(/\n$/, '').split('\n');
  const line = lines[lines.length - 1];
  assert.ok(line.startsWith('sig\ted25519\t') && line.length === 'sig\ted25519\t'.length + 88 && signed.at(-1) === 0x0a);
  // Plain Ed25519 over the plan's bytes, checked without this module.
  const sig = Buffer.from(line.slice(12), 'base64');
  assert.ok(crypto.verify(null, Buffer.from(PLAN), publicKeyFromRaw(s.pub), sig));
  assert.equal(verifyFor(s.pub, signed, 'tjfq5rqwnnrxk3m9q2x7v4p8ab').toString(), PLAN);
  // CRLF on the signature line, and a missing final newline, are tolerated.
  verify(s.pub, Buffer.concat([signed.subarray(0, -1), Buffer.from('\r\n')]));
  verify(s.pub, signed.subarray(0, -1));
  assert.throws(() => s.sign(signed), /already signed/);
  assert.throws(() => s.sign(Buffer.from('ti-record\t1\n')), /not a ti-plan/);
});

test('no final newline: it is added, and signed', (t) => {
  const { s } = newSigner(t);
  const signed = s.sign(Buffer.from(PLAN.replace(/\n$/, '')));
  assert.ok(signed.subarray(0, PLAN.length).equals(Buffer.from(PLAN)));
  verify(s.pub, signed);
});

test('rejects', (t) => {
  const { s } = newSigner(t);
  const { s: other } = newSigner(t);
  const signed = s.sign(Buffer.from(PLAN)).toString('latin1');
  const flip = (() => { const i = signed.lastIndexOf('\t') + 5; return signed.slice(0, i) + (signed[i] === 'A' ? 'B' : 'A') + signed.slice(i + 1); })();
  const cases = {
    'tampered url': signed.replace('echo hi', 'echo HI'),
    'dropped line': signed.replace('name\tHello\n', ''),
    'added line': signed.replace('[target]\n', '[target]\nstep\trun\tevil\n'),
    'text after sig': signed + 'launch\tevil\n',
    'blank after sig': signed + '\n',
    'flipped sig bit': flip,
  };
  for (const [name, doc] of Object.entries(cases)) assert.throws(() => verify(s.pub, Buffer.from(doc, 'latin1')), VerifyError, name);
  assert.throws(() => verify(other.pub, Buffer.from(signed, 'latin1')), /does not verify/);
  assert.throws(() => verify(s.pub, Buffer.from(PLAN)), /not signed/);
  assert.throws(() => verifyFor(s.pub, Buffer.from(signed, 'latin1'), 'a'.repeat(26)), /is for record/);
  // A validly signed plan for another record is refused.
  const p2 = s.sign(Buffer.from(PLAN.replace('tjfq5rqwnnrxk3m9q2x7v4p8ab', 'b'.repeat(26))));
  assert.throws(() => verifyFor(s.pub, p2, 'tjfq5rqwnnrxk3m9q2x7v4p8ab'));
  // Non-canonical S (S + L) is refused.
  const i = signed.lastIndexOf('\t') + 1;
  const sig = Buffer.from(signed.slice(i, -1), 'base64');
  const L = Buffer.from('edd3f55c1a631258d69cf7a2def9de1400000000000000000000000000000010', 'hex');
  let c = 0;
  for (let k = 0; k < 32; k++) { const v = sig[32 + k] + L[k] + c; sig[32 + k] = v & 255; c = v >> 8; }
  assert.throws(() => verify(s.pub, Buffer.from(signed.slice(0, i) + sig.toString('base64') + '\n', 'latin1')));
});

test('recordOf and split', () => {
  assert.equal(recordOf(PLAN), 'tjfq5rqwnnrxk3m9q2x7v4p8ab');
  assert.equal(recordOf('ti-plan\t1\n\n[target]\nrecord\tx\n'), '', 'a record line inside a target block is not the header\'s');
  assert.equal(split(Buffer.from(PLAN)).ok, false);
});

// build/sign_test.go
test('addRequestLine', (t) => {
  const plan = 'ti-plan\t1\nrecord\tabc\n\n[target]\nwhen\tlinux\t0\t9999\t*\n';
  assert.equal(addRequestLine(plan, 'name', 'python', 'Some.Pkg'), 'ti-plan\t1\nrequest\tname\tpython\tSome.Pkg\nrecord\tabc\n\n[target]\nwhen\tlinux\t0\t9999\t*\n');
  // Tabs and newlines in a value can't add lines or fields.
  const got = addRequestLine(plan, 'name', 'python', 'a\tb\nlaunch\tevil');
  assert.equal(got.split('\n').length, plan.split('\n').length + 1);
  assert.throws(() => addRequestLine('ti-record\t1\n', 'name'));
  // The line is inside the signature.
  const { s } = newSigner(t);
  const signed = s.signString(addRequestLine(plan, 'name', 'python', 'requests'));
  assert.throws(() => verify(s.pub, Buffer.from(signed.replace('requests', 'evil'))));
});

// A second request line goes after the first, so `name` stays the first
// `request` line an engine reads (docs/format.md section 1).
test('addRequestLine keeps the order: name, then nonce', () => {
  const plan = 'ti-plan\t1\nrecord\tabc\n\n[target]\nwhen\tlinux\t0\t9999\t*\n';
  const hex = 'a'.repeat(32);
  const both = addRequestLine(addRequestLine(plan, 'name', 'python', 'requests'), 'nonce', hex);
  assert.equal(both.split('\n').slice(0, 4).join('\n'),
    'ti-plan\t1\nrequest\tname\tpython\trequests\nrequest\tnonce\t' + hex + '\nrecord\tabc');
  // A nonce alone is the second line, as `name` would have been.
  assert.equal(addRequestLine(plan, 'nonce', hex).split('\n')[1], 'request\tnonce\t' + hex);
});

// The revocation list is signed with the same key by the same rules, and
// its header keeps the two kinds of document apart (docs/format.md 7).
test('signing another kind of document', (t) => {
  const { s } = newSigner(t);
  const doc = 'ti-revocations\t1\nissued\t2026-09-20T11:00:00Z\nrevoke\trecord\tabc\n';
  const signed = s.signStringAs('ti-revocations', doc);
  assert.equal(verify(s.pub, Buffer.from(signed), 'ti-revocations').toString('utf8'), doc);
  // Not as a plan, and a plan is not one of these.
  assert.throws(() => verify(s.pub, Buffer.from(signed)), VerifyError);
  assert.throws(() => verify(s.pub, Buffer.from(s.signString(PLAN)), 'ti-revocations'), VerifyError);
  assert.throws(() => s.signStringAs('ti-revocations', PLAN));
  // A changed entry does not verify.
  assert.throws(() => verify(s.pub, Buffer.from(signed.replace('abc', 'xyz')), 'ti-revocations'), VerifyError);
});
